File: java/src/ch1/ex1_7.java
package ch1;

import java.util.function.Function;

public class ex1_7 {
    static class SquareRootFunction {
        static final double factor = 100000.0;

        double calc(double x) {
            return sqrt_iter(1.0, x);
        }

        double sqrt_iter(double guess, double x) {
            return is_good_enough(guess, x) ? guess : sqrt_iter(improve(guess, x), x);
        }

        boolean is_good_enough(double guess, double x) {
            return Math.abs(square(guess) - x) < (guess / factor);
        }

        double square(double x) {
            return x * x;
        }

        double improve(double guess, double x) {
            return average(guess, x / guess);
        }

        double average(double x, double y) {
            return (x + y) / 2.0;
        }
    }

    static class sqrt extends SquareRootFunction implements Function<Double, Double> {
        @Override
        public Double apply(Double x) {
            return calc(x);
        }
    }

    public static void main(String[] args) {
        System.out.println(new sqrt().apply(0.02) + " vs. actual\n" + Math.sqrt(0.02));
        System.out.println(new sqrt().apply(0.0002) + " vs. actual\n" + Math.sqrt(0.0002));
    }
}

File: python/ch1/ex1-7.py
import unittest
import math


def sqrt(x):
    return sqrt_iter(1.0, x)


def sqrt_iter(guess, x):
    return guess if is_good_enough(guess, x) else sqrt_iter(improve(guess, x), x)


def improve(guess, x):
    return average(guess, x / guess)


def average(x, y):
    return (x + y) / 2.0


def is_good_enough(guess, x):
    return abs(square(guess) - x) < (guess / 100000)


def square(x):
    return x * x


# ------------------------------

class MyTest(unittest.TestCase):

    def test_sqrt(self):
        self.assertAlmostEqual(sqrt(0.02), math.sqrt(0.02), places=6)
        self.assertAlmostEqual(sqrt(0.0002), math.sqrt(0.0002), places=6)



if __name__ == '__main__':
    unittest.main()
